sentry-android-replay/src/main/java/io/sentry/android/replay/ReplayIntegration.kt

```diff
@@ -0,0 +1,317 @@
+package io.sentry.android.replay
+
+import android.content.Context
+import android.graphics.Bitmap
+import android.graphics.Point
+import android.graphics.Rect
+import android.os.Build
+import android.os.Build.VERSION
+import android.os.Build.VERSION_CODES
+import android.view.WindowManager
+import io.sentry.DateUtils
+import io.sentry.Hint
+import io.sentry.IHub
+import io.sentry.Integration
+import io.sentry.ReplayRecording
+import io.sentry.SentryLevel.DEBUG
+import io.sentry.SentryLevel.INFO
+import io.sentry.SentryOptions
+import io.sentry.SentryReplayEvent
+import io.sentry.protocol.SentryId
+import io.sentry.rrweb.RRWebMetaEvent
+import io.sentry.rrweb.RRWebVideoEvent
+import io.sentry.transport.ICurrentDateProvider
+import io.sentry.util.FileUtils
+import java.io.Closeable
+import java.io.File
+import java.util.Date
+import java.util.concurrent.ExecutorService
+import java.util.concurrent.Executors
+import java.util.concurrent.ThreadFactory
+import java.util.concurrent.TimeUnit.MILLISECONDS
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.atomic.AtomicReference
+import kotlin.LazyThreadSafetyMode.NONE
+import kotlin.math.roundToInt
+
+class ReplayIntegration(
```

Comment:
Alternatively to exposing ReplayCache to Flutter, it may be even better to expose the functionality in this class, but with custom actual `recorder` implementation in Flutter

Reply:
I think I'd opt for exposing `ReplayCache` and some methods of this class (like `captureReplay`) through `InternalSentrySdk`, wdyt?

then you'd soemthing like:
```kotlin
val cache = ReplayCache(options, replayId, ScreenshotRecorderConfig(width, height, frameRate, bitRate))
cache.addFrame(screenshotFilename)

val video = cache.createVideoOf(duration, timestamp, segmentId)
// either full envelope details
val replayData = mapOf("replay_event" to mapOf(...), "replay_recording" to mapOf(...), "video" to videoFilePath)
InternalSentrySDK.captureReplay(replayData)

// or just the params we need for replay, and we build the event on our side
val replayData = mapOf(
        video: File,
        currentReplayId: SentryId,
        segmentTimestamp: Date,
        segmentId: Int,
        frameCount: Int,
        duration: Long,
        replayType: ReplayType
)
InternalSentrySDK.captureReplay(replayData)
```

Reply:
looks good to me with either API

Reply:
probably the former is better cause you have full control of what to send then, and wouldn't need to update the native sdk everytime you want to change/add something

Reply:
I understand. As for the beforeSend callback. Does the event already have the binary data loaded or is it still a file path in the event?

Reply:
still a file path, we load it later here https://github.com/getsentry/sentry-java/blob/79151e9a7103dd1f1f945648a22c05d1554e4fd5/sentry/src/main/java/io/sentry/SentryEnvelopeItem.java#L377-L385

Reply:
that sounds great then!

Reply:
@vaind lemme know if this works: https://github.com/getsentry/sentry-java/pull/3272/commits/0f4e718fc20c2001efed3ac4ec1f1947f4366a46

Reply:
looks good afaict without testing this out, which I'm not sure I'd know how to include in a flutter app without it being released.

Maybe it's worth adding a test case in this repo to check the workflow with image files? It would also prevent future regressions.

Reply:
yeah, I'll add a test 👍 you can absolutely test it from a branch by using gradle included builds, that's how I test it now on a pocketcasts app

in `settings.gradle(.kts)`:
```
includeBuild("path/to/sentry-java") {
  dependencySubstitution {
    substitute(module("io.sentry:sentry")).using(project(":sentry"))
    substitute(module("io.sentry:sentry-android-core")).using(project(":sentry-android-core"))
    substitute(module("io.sentry:sentry-replay")).using(project(":sentry-android-replay"))
  }
}
```

you just have to add a fake dependency `io.sentry:sentry-replay:1.0.0` alongside `sentry-android:x.y.z` until it's shipped to maven

Reply:
added test in https://github.com/getsentry/sentry-java/pull/3272/commits/18eb67ebf6a7ea188175fb819fe67c7a76d7a561